nimbus/src/lib.rs

```diff
@@ -193,16 +237,31 @@ impl NimbusClient {
     // (Useful for testing so you can have some control over what experiments
     // are enrolled)
     pub fn set_nimbus_id(&self, uuid: &Uuid) -> Result<()> {
-        let db = self.db()?;
+        let db = self.db()?.lock().unwrap();
         let mut writer = db.write()?;
         db.get_store(StoreId::Meta)
             .put(&mut writer, DB_KEY_NIMBUS_ID, uuid)?;
         writer.commit()?;
         Ok(())
     }
 
-    fn db(&self) -> Result<&Database> {
-        self.db.get_or_try_init(|| Database::new(&self.db_path))
+    fn db(&self) -> Result<&Mutex<Database>> {
+        self.db
+            .get_or_try_init(|| Ok(Mutex::new(Database::new(&self.db_path)?)))
+    }
+
+    // Call this function whenever it's possible that our enrollments might
+    // have changed. It should be held while the database mutex is held, but
+    // after all IO is complete.
+    fn update_cached_experiment_branches(&self, db: &Database) -> Result<()> {
+        let cell = self.cached_experiment_branches.write().unwrap();
+        let experiments = get_enrollments(&db)?;
```

Comment:
IIUC this is doing to hold the cache write lock for the duration of a DB read, which means that *technically* a concurrent call to `get_experiment_branch` could still have to wait for `get_enrollments` to complete. Would it still work if we read the experiments first, and took the write lock *only* for the duration of the `cell.replace()`? I think the fact that we're holding the database mutex means there won't be any concurrent attempts to update the cache, so there's no risk of writing stale data.

Reply:
Nice catch, thanks!